service/history/shard/controller_impl.go

```diff
@@ -446,12 +464,63 @@ func (c *ControllerImpl) acquireShards(ctx context.Context) {
 	_ = sem.Acquire(ctx, concurrency)
 
 	c.RLock()
+
 	numOfOwnedShards := len(c.historyShards)
+	// Readiness check: We should set initialShardsAcquired when:
+	// 1. It's not already set.
+	// 2. We should own at least one shard (i.e. not before we join membership).
+	// 3. We have ownership of all the shards we're supposed to own.
+	if !c.initialShardsAcquired.Ready() && numOfOwnedShards > 0 {
+		go c.checkShardReadiness(readinessCtx, readinessCancel, expmaps.Values(c.historyShards))
```

Comment:
Perhaps we could avoid passing readinessCancel and rewrite to:

```
go func() {
    defer readinessCancel()
    c.checkShardReadiness(readinessCtx, readinessCancel, expmaps.Values(c.historyShards))
}()
```

That way, both branches of this predicate clearly handle canceling the readiness context.

Reply:
sure, that's nice